libs/Facilitator/Facilitator.js

```diff
@@ -0,0 +1,387 @@
+      gasLimit.toString(10),
+      this.stakerNonce.toString(10),
+      this.hashLockObj.hashLock
+    );
+
+    return this.sendTransaction(tx, transactionOptions);
+  }
+
+  /**
+   * @function getGatewayNonce
+   *
+   * Returns the gateway nonce for the given account address.
+   *
+   * @param {string} accountAddress Account address for which the nonce is to be fetched.
+   *
+   * @returns {Promise} Promise object represents the nonce of account address.
+   */
+  async getGatewayNonce(accountAddress) {
+    if (!web3.utils.isAddress(accountAddress)) {
+      throw new Error('Invalid account address.');
+    }
+
+    const gatewayContract = this.contracts.Gateway(this.gatewayAddress);
+
+    return gatewayContract.methods
+      .getNonce(accountAddress)
+      .call()
+      .then((nonce) => {
+        console.log(`\t - Gateway Nonce for ${accountAddress}:`, nonce);
+        return nonce;
+      });
+  }
+
+  /**
+   * @function getBounty
+   *
+   * Returns the bounty amount from EIP20Gateway contract.
+   *
+   * @returns {Promise} Promise object represents the bounty.
+   */
+  getBounty() {
+    const gatewayContract = this.contracts.Gateway(this.gatewayAddress);
+
+    return gatewayContract.methods
+      .bounty()
+      .call()
+      .then((bounty) => {
+        console.log('\t - Gateway Bounty:', bounty);
+        return bounty;
+      });
+  }
+
+  /**
+   * @function getBaseToken
+   *
+   * Returns the ERC20 base token address.
+   *
+   * @returns {Promise} Promise object represents ERC20 base token address.
+   */
+  getBaseToken() {
+    const oThis = this;
+    if (oThis.baseTokenAddress) {
+      return oThis.baseTokenAddress;
+    }
+    const gatewayContract = this.contracts.Gateway(this.gatewayAddress);
+
+    return gatewayContract.methods
+      .baseToken()
+      .call()
+      .then((baseToken) => {
+        console.log('\t - Gateway baseToken:', baseToken);
+        oThis.baseTokenAddress = baseToken;
+        return baseToken;
+      });
+  }
+
+  /**
+   * @function getValueToken
+   *
+   * Returns the EIP20 token address.
+   *
+   * @returns {Promise} Promise object represents EIP20 token address.
+   */
+  getValueToken() {
+    const oThis = this;
+    if (oThis.valueTokenAddress) {
+      return oThis.valueTokenAddress;
+    }
+    const gatewayContract = this.contracts.Gateway(this.gatewayAddress);
+
+    return gatewayContract.methods
+      .token()
+      .call()
+      .then((token) => {
+        console.log('\t - Gateway token:', token);
+        oThis.valueTokenAddress = token;
+        return token;
+      });
+  }
+
+  /**
+   * @function approveBountyAmount
+   *
+   * Approves gateway address for the bounty amount transfer.
+   *
+   * @param {string} facilitator Facilitator address.
+   * @param {string} gas The gas provided for the transaction execution
+   *
+   * @returns {Promise} Promise object.
+   */
+  async approveBountyAmount(facilitator, gas = '7000000') {
```

Comment:
I think we should check if the current value of `allowance` is smaller than `bounty` before doing a `approve` transaction, as the `approve` might actually lower the `allowance`. A common optimization for workers I've seen is that you approve a much bigger pool of of tokens up front and then use that allowance up over time, which is not possible if we set it to the conrete amount every time.

Same for `approveStakeAmount`.